packages/backend/src/app-connection/app-connection-service.ts

```diff
@@ -0,0 +1,162 @@
+    switch (connection.value.type) {
+        case AppConnectionType.CLOUD_OAUTH2:
+            connection.value = await refreshCloud(connection.appName, connection.value);
+            break;
+        case AppConnectionType.OAUTH2:
+            connection.value = await refreshWithCredentials(connection.value);
+            break;
+        case AppConnectionType.CUSTOM:
+            for (const key in Object.keys(connection.value)) {
+                let connectionValue = connection.value[key];
+                if (typeof connectionValue === 'object' && connectionValue.hasOwnProperty('type')) {
+                    let type: AppConnectionType = connectionValue.type;
+                    switch (type) {
+                        case AppConnectionType.CLOUD_OAUTH2:
+                            connectionValue = await refreshCloud(connection.appName, connectionValue as CloudOAuth2ConnectionValue);
+                            break;
+                        case AppConnectionType.OAUTH2:
+                            connectionValue = await refreshWithCredentials(connectionValue as OAuth2ConnectionValueWithApp);
+                            break;
+                        default:
+                            break;
+                    }
+                }
+            }
+        default:
+            break;
+    }
+    return connection;
+}
+
+const REFRESH_THRESHOLD = 15 * 60; // Refresh if there is less than 15 minutes to expire
+
+function isExpired(connection: BaseOAuth2ConnectionValue) {
+    const secondsSinceEpoch = Math.round(Date.now() / 1000);
+
+    if (!connection.expires_in || !connection.refresh_token) {
+        return false;
+    }
+
+    return (connection.claimed_at + connection.expires_in + REFRESH_THRESHOLD <= secondsSinceEpoch)
+}
+
+async function refreshCloud(appName: string, connectionValue: CloudOAuth2ConnectionValue): Promise<CloudOAuth2ConnectionValue> {
+    if (!isExpired(connectionValue)) {
+        return connectionValue;
+    }
+
+    const response = (
+        await axios.post("https://secrets.activepieces.com/refresh", {
+            refreshToken: connectionValue.refresh_token,
+            pieceName: appName,
+        } as RefreshTokenFromCloudRequest)
+    ).data;;
+
+    return {
+        ...response,
+        type: AppConnectionType.CLOUD_OAUTH2
+    }
+}
+
+async function refreshWithCredentials(appConnection: OAuth2ConnectionValueWithApp): Promise<OAuth2ConnectionValueWithApp> {
+    if (!isExpired(appConnection)) {
+        return appConnection;
+    }
+    const settings = appConnection;
+    const response = (
+        await axios.post(
+            settings.token_url,
+            qs.stringify({
+                client_id: settings.client_id,
+                client_secret: settings.client_secret,
+                redirect_uri: settings.redirect_url,
+                grant_type: "refresh_token",
+                refresh_token: appConnection.refresh_token,
+            }),
```

Comment:
FYI, `qs` is deprecated, `URLSearchParams` is the new API.

Reply:
Removed thanks